### ISSUE TITLE
Hard to get a green build because of the json output comparison

### ISSUE THREAD

**Issue by tisto** [plone/plone.restapi#263]
I am currently trying to get a trivial fix into plone.restapi and the json output comparison drives me crazy. I get lots of those errors:

1) Upper/lowercase mix of headers (RFC2616 says HTTP headers are case-insensitive: https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2)

-content-type: application/json
+Content-Type: application/json

e.g. https://travis-ci.org/plone/plone.restapi/jobs/210535550

2) Ordering mixture:

 GET /plone/non-existing-resource HTTP/1.1
-Authorization: Basic YWRtaW46c2VjcmV0
 Accept: application/json
+Authorization: Basic YWRtaW46c2VjcmV0

e.g. https://travis-ci.org/plone/plone.restapi/jobs/210493416

We either have to force an ordering and lower/uppercase or disable this check for now. cc @lukasgraf @datakurre

**Comment by tisto** [comment]
Another popular issue: new line at the end of the file. No idea where this comes from:

```
@@ -1,9 +1,9 @@
 HTTP/1.1 200 OK
-Content-Type: application/json
+content-type: application/json
 
 [
   {
     "source": "http://localhost:55001/plone/front-page", 
     "target": "http://localhost:55001/plone/folder/front-page"
   }
-]
+]
\ No newline at end of file
```

**Comment by tisto** [comment]
I added a fix that uses an ordered dict for the headers and forces lower-case on the HTTP header keys:

https://travis-ci.org/plone/plone.restapi/builds/210547633

This does not solve all issues, but most of them. Question would be if we prefer upper- or lowercase HTTP header keys. I guess we could try to use Python's title() method to force first-letter-uppercase.

**Comment by tisto** [comment]
Both lowercase:

https://github.com/plone/plone.restapi/pull/264

and uppercase HTTP headers work:

https://github.com/plone/plone.restapi/pull/265

**Comment by tisto** [comment]
Opinions on lower- vs uppercase here? @lukasgraf @datakurre @sneridagh @buchi ?

**Comment by sneridagh** [comment]
@tisto my five cents here, although I think that it's a matter of taste: Uppercase.

**Comment by lukasgraf** [comment]
@tisto sorry, only stumbled upon this now. Happend to have a badly timed MS relapse and writing from the hospital right now. Might be off for a few days.

TLDR:
- (Strong) preference towards `Camel-Case` headers
- 👍 for dumping headers using ordered dict, and writing them in a case-canonicalized way
- whitespace issues might be `git` related, and may need to be solved in test that diffs, so it diffs whitespace-agnostic

---

I agree with the whitespace / newline issue: That most likely is related to `git` - git has various settings regarding whitespace handling, which also are somehwat git-version dependent, including options that control whether git actually shows whitespace-only differences in `git diff` or not, and whether it is supposed to change whitespace (canonicalize it according to the local user's preferences) when checking in content.

I could see how those settings on peoples machines could lead to failures in my test, if whitepsace isn't exactly the same, and we don't really control the settings of `git` on Travis. -> Should fix my test to only fail if changes aren't whitespace-only differences.

---

HTTP header order is obvious, if that's not controlled when storing (preferable) or compared in a order-independent way in the test, that's an issue.

Header-Casing seems to be flaky in terms of that the `requests` module doesn't always seem to reliably produce one or the other style in responses (maybe also requests), from what I can tell. 👍 For canonicalizing them ourselves using `str.title()` (instead of making the actual test case-insensitive).

Sorry for all the fallout from this 😭

**Comment by tisto** [comment]
@lukasgraf sorry to hear. Hope you get well soon! Don't worry about Plone stuff. Your health is far more important than anything else!

I merged the Camel-Case header option. The problem is solved for now I guess. I rarely saw the whitespace issue, so I guess we can live with it for now. Thank you for writing the docs test in the first place! The scripted already helped me quite a few times when I forgot to update/commit the docs.

**Comment by tisto** [comment]
I will close the issue for now because the problem is solved for now. We can open a new issue if it should pop up again.
